### PR TITLE
PW-1638: Integrate OneClick Payments

### PR DESCRIPTION
**Pull request by Aleffio**

### CODE REVIEW

File: controllers/front/Payment.php [hunk 1]
```diff
@@ -400,4 +404,19 @@ public function buildBrowserData($request = [])
...
     }
 
+    /**
+     * @param array $request
+     * @param $payload
+     * @return array
+     */
+    public function buildRecurringData($request = [], $payload)
+    {
+        if (!empty($payload['storeCc']) && strpos($payload['storeCc'], 'true') !== false) {
```
Comment (cyattilakiss) [inline controllers/front/Payment.php]: why checking it with strpos? Can we do a simple === ?

File: helper/Data.php [hunk 2]
```diff
@@ -100,6 +108,42 @@ public function getOriginKeyForOrigin()
...
+                ),
+            ],
+            "shopperReference" => $shopperReference,
+            "shopperLocale" => $shopperLocale
+        ];
+
+        try {
+            $responseData = $this->adyenCheckoutService->paymentMethods($adyFields);
+        } catch (\Adyen\AdyenException $e) {
+            $this->helperData->adyenLogger()->logError("There was an error retrieving the payment methods. message: " . $e->getMessage());
```
Comment (cyattilakiss) [inline helper/Data.php]: here we don't have responseData and the return will fail

File: views/templates/front/adyencheckout.tpl [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{if !$originKey}
+    <h5>There is an error with retrieving the originKey,
+        please check your API key in the Adyen Module configuration</h5>
+{else}
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
+    <script>
+        $(document).ready(function () {
+            window.adyenCheckout = new AdyenCheckout({
```
Comment (cyattilakiss) [inline views/templates/front/adyencheckout.tpl]: can we also pass here the paymentMethods response so we don't have to do it everyime for the other components?

File: views/templates/front/payment.tpl [hunk 4]
```diff
@@ -17,8 +24,11 @@
                     <script type="text/javascript" src="/js/jquery/plugins/fancybox/jquery.fancybox.js"></script>
                 {/if}
 
+
                 <script>
 
+                    $(document).ready(function () {
+                    // window.onload = function(){
```
Comment (cyattilakiss) [inline views/templates/front/payment.tpl]: remove if not necessary

File: adyen.php [hunk 5]
```diff
@@ -547,19 +549,64 @@ public function hookOrderConfirmation($params)
...
+        $amount = $this->context->cart->getOrderTotal();
+        $currency = $this->context->currency->iso_code;
+        $address = new Address($this->context->cart->id_address_invoice);
+        $isoAddress = Country::getIsoById($address->id_country);
+        $shopperReference = $this->context->cart->id_customer;
+        $shopperLocale = $this->context->language->iso_code;
+
+        //retrieve payment methods
+        $paymentMethods = $this->helper_data->fetchPaymentMethods($isoAddress, $amount, $currency, $shopperReference, $shopperLocale);
+        if(!empty($paymentMethods) && isset($paymentMethods['oneClickPaymentMethods'])){
```
Comment (rikterbeek) [inline adyen.php]: can't we simplify this to if(!empty($paymentMethods['oneClickPaymentMethods']))

File: adyen.php [hunk 6]
```diff
@@ -547,19 +549,64 @@ public function hookOrderConfirmation($params)
...
+            $oneClickPaymentMethods = $paymentMethods['oneClickPaymentMethods'];
+        }
+
         $payment_options = array();
+
+        if(isset($oneClickPaymentMethods)) {
```
Comment (rikterbeek) [inline adyen.php]: same check as above right can't we move logic into that if statement.

File: adyen.php [hunk 7]
```diff
@@ -547,19 +549,64 @@ public function hookOrderConfirmation($params)
...
+            foreach ($oneClickPaymentMethods as $storedCard) {
+                if (isset($storedCard["storedDetails"]["card"]["expiryMonth"])) {
```
Comment (rikterbeek) [inline adyen.php]: better to use !empty

File: adyen.php [hunk 8]
```diff
@@ -547,19 +549,64 @@ public function hookOrderConfirmation($params)
      */
     public function hookPaymentOptions($params)
     {
+        $amount = $this->context->cart->getOrderTotal();
+        $currency = $this->context->currency->iso_code;
+        $address = new Address($this->context->cart->id_address_invoice);
+        $isoAddress = Country::getIsoById($address->id_country);
+        $shopperReference = $this->context->cart->id_customer;
+        $shopperLocale = $this->context->language->iso_code;
```
Comment (rikterbeek) [inline adyen.php]: below we use $this->helper_data->getLocale($this->context) should we use it here as well ?

File: adyen.php [hunk 9]
```diff
@@ -585,19 +632,58 @@ public function hookPayment($params)
...
             array(
-                'locale' => $this->context->language->iso_code, // no locale in Prestashop1.6 only languageCode that is en-en but we need en_EN
+                'locale' => $this->helper_data->getLocale($this->context), // no locale in Prestashop1.6 only languageCode that is en-en but we need en_EN
                 'originKey' => $this->helper_data->getOriginKeyForOrigin(),
                 'environment' => \Configuration::get('ADYEN_MODE'),
                 'paymentProcessUrl' => $this->context->link->getModuleLink($this->name, 'Payment', array(), true),
                 'threeDSProcessUrl' => $this->context->link->getModuleLink($this->name, 'ThreeDSProcess', array(), true),
-                'prestashop16' => true
+                'prestashop16' => true,
+                'oneClickPaymentMethod' => "",
```
Comment (rikterbeek) [inline adyen.php]: why do we define an empty string  ?

File: views/templates/front/payment.tpl [hunk 10]
```diff
@@ -33,12 +43,10 @@
                     let language;
                     let javaEnabled;
 
-                    let adyenCheckout;
-
                     let placeOrderAllowed;
                     let popupModal;
 
-
+                    let storeCc;
```
Comment (rikterbeek) [inline views/templates/front/payment.tpl]: @KadoBOT can you validate if this javascript will work for all common browsers ? I think better to move from let to var.

File: views/templates/front/payment.tpl [hunk 11]
```diff
@@ -4,6 +4,13 @@
         please check your API key in the Adyen Module configuration</h5>
 </form>
 {else}
+    {if $prestashop16}
+        <p></p>
+        <div class="adyen-payment-method-label">
+            Pay with Credit Card
```
Comment (rikterbeek) [inline views/templates/front/payment.tpl]: Can't we add smarty translation here so merchants can change label ?

File: views/templates/front/adyencheckout.tpl [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+                locale: "{$locale}",
+                originKey: "{$originKey}",
+                environment: "{$environment}",
+                risk: {
+                    enabled: false
```
Comment (rikterbeek) [inline views/templates/front/adyencheckout.tpl]: let's remove this and validate if it still work. This was added for a quickfix that should be solved by now.

File: views/templates/front/adyencheckout.tpl (hunk 3 above)
Comment (Aleffio) [reply views/templates/front/adyencheckout.tpl]: it is on another template at the moment, will find a way to not duplicate paymentMethods response

File: helper/Data.php [hunk 13]
```diff
@@ -100,6 +108,43 @@ public function getOriginKeyForOrigin()
...
+     * @param $store
+     * @param $country
+     * @return array
+     */
+    public function fetchPaymentMethods($countryCode, $amount, $currency, $shopperReference, $shopperLocale)
+    {
+        $merchantAccount = \Configuration::get('ADYEN_MERCHANT_ACCOUNT');
+
+        if (!$merchantAccount) {
+            return [];
```
Comment (cyattilakiss) [inline helper/Data.php]: would be nice to show or log a message that the merchant account is not configured therefore no payment methods are retrieved.

File: adyen.php [hunk 14]
```diff
@@ -606,21 +653,55 @@ public function hookPayment()
 
         $this->context->controller->addCSS($this->_path . 'css/adyen.css', 'all');
 
+        $amount = $this->context->cart->getOrderTotal();
+        $currency = $this->context->currency->iso_code;
```
Comment (msilvagarcia) [inline adyen.php]: You should get the currency from the cart object:
```php
$currencyData = Currency::getCurrency($cart->id_currency);
$currency = $currencyData['iso_code'];
```

File: adyen.php [hunk 15]
```diff
@@ -569,25 +570,71 @@ public function hookOrderConfirmation()
     public function hookPaymentOptions()
     {
         $payment_options = array();
+        if(!$this->context->customer->is_guest) {
+            $amount = $this->context->cart->getOrderTotal();
+            $currency = $this->context->currency->iso_code;
```
Comment (msilvagarcia) [inline adyen.php]: You should get the currency from the cart object:
```php
$currencyData = Currency::getCurrency($cart->id_currency);
$currency = $currencyData['iso_code'];
```

File: adyen.php [hunk 16]
```diff
@@ -569,25 +570,71 @@ public function hookOrderConfirmation()
...
+            $shopperLocale = $this->helper_data->getLocale($this->context);
+
+
+            //retrieve payment methods
+            $paymentMethods = $this->helper_data->fetchPaymentMethods($isoAddress, $amount, $currency, $shopperReference, $shopperLocale);
+
+            if (!empty($paymentMethods['oneClickPaymentMethods'])) {
+                $oneClickPaymentMethods = $paymentMethods['oneClickPaymentMethods'];
+                foreach ($oneClickPaymentMethods as $storedCard) {
+                    if (!empty($storedCard["storedDetails"]["card"]["expiryMonth"])) {
```
Comment (msilvagarcia) [inline adyen.php]: Why is this checking for expiry month? Can we make the reason more obvious somehow?

File: adyen.php [hunk 17]
```diff
@@ -606,21 +653,55 @@ public function hookPayment()
...
+                            'oneClickPaymentMethod' => json_encode($storedCard),
+                            'recurringDetailReference' => $storedCard['recurringDetailReference'],
+                            'name' => $storedCard['name'],
+                            'number' => $storedCard['storedDetails']['card']['number']
+                        )
+                    );
+                }
+                $payments .= $this->display(__FILE__, '/views/templates/front/oneclick.tpl');
+            }
+        }
```
Comment (msilvagarcia) [inline adyen.php]: This block doesn't look too different from the previous one. Can we extract a method to unify these?

File: controllers/front/Payment.php [hunk 18]
```diff
@@ -393,4 +397,19 @@ public function buildBrowserData($request = [])
         return $request;
     }
 
+    /**
+     * @param array $request
+     * @param $payload
+     * @return array
+     */
+    public function buildRecurringData($request = [], $payload)
```
Comment (msilvagarcia) [inline controllers/front/Payment.php]: Short array syntax won't work in PHP 5.2

File: helper/Data.php [hunk 19]
```diff
@@ -100,6 +108,43 @@ public function getOriginKeyForOrigin()
         return $originKey;
     }
 
+    /**
+     * @param $store
+     * @param $country
+     * @return array
+     */
+    public function fetchPaymentMethods($countryCode, $amount, $currency, $shopperReference, $shopperLocale)
```
Comment (msilvagarcia) [inline helper/Data.php]: Most of these parameters are in the Cart object. Passing the Cart object here would make calling this object less complex.

File: helper/Data.php [hunk 20]
```diff
@@ -100,6 +108,43 @@ public function getOriginKeyForOrigin()
...
+        ];
+
+        $responseData = "";
+        try {
+            $responseData = $this->adyenCheckoutService->paymentMethods($adyFields);
+        } catch (\Adyen\AdyenException $e) {
+            $this->helperData->adyenLogger()->logError("There was an error retrieving the payment methods. message: " . $e->getMessage());
+        }
+        return $responseData;
+    }
```
Comment (msilvagarcia) [inline helper/Data.php]: This method would be better in a specialized object

File: helper/Data.php [hunk 21]
```diff
@@ -375,6 +420,20 @@ public function startSession()
...
+    public function getLocale($context)
+    {
+        // no locale in Prestashop1.6 only languageCode that is en-en but we need en_EN
+        if ($this->isPrestashop16()) {
+            return $context->language->iso_code;
+        } else {
+            return $context->language->locale;
+        }
+    }
+
```
Comment (msilvagarcia) [inline helper/Data.php]: This method would be better in a specialized object

File: views/templates/front/payment.tpl (hunk 10 above)
Comment (msilvagarcia) [reply views/templates/front/payment.tpl]: https://caniuse.com/#feat=let
Doesn't work until IE 10

File: controllers/front/Payment.php [hunk 22]
```diff
@@ -286,6 +287,17 @@ public function buildCCData($request, $payload)
...
 
+        $shopperReference = $this->context->cart->id_customer;
+        if(!empty($shopperReference)) {
+            $request['shopperReference'] = $shopperReference;
+        }
+
+        //Oneclick data
+        if(!empty($payload['recurringDetailReference']) &&
+        $recurringDetailReference = $payload['recurringDetailReference']) {
+            $request['paymentMethod']['recurringDetailReference'] = $recurringDetailReference;
```
Comment (cyattilakiss) [inline controllers/front/Payment.php]:
```suggestion
            $request['paymentMethod']['recurringDetailReference'] = $payload['recurringDetailReference'];
```

Bit easier to read IMO but you can keep as it is if you think :)

File: controllers/front/Payment.php [hunk 23]
```diff
@@ -286,6 +287,17 @@ public function buildCCData($request, $payload)
...
         }
 
+        $shopperReference = $this->context->cart->id_customer;
+        if(!empty($shopperReference)) {
+            $request['shopperReference'] = $shopperReference;
+        }
+
+        //Oneclick data
+        if(!empty($payload['recurringDetailReference']) &&
+        $recurringDetailReference = $payload['recurringDetailReference']) {
```
[the remark this answers is not in the document]
Comment (cyattilakiss) [reply controllers/front/Payment.php]: or maybe the whole recheck is not even important since the !empty part will check these so the `&&
        $recurringDetailReference = $payload['recurringDetailReference']` can just be removed
